test(ivy): test against recent Material commit

JoostK opened the pull request:
Updating material-unit-tests to run against master with Bazel

JoostK committed:
test(ivy): update Material to recent commit from master branch

Previously, the ivy-2019 branch of the Material (aka components) repo was
used, which contains some changes that were necessary to work with Ivy.
These changes are not longer necessary, as Material's master branch is
fully working with Ivy today. To be up-to-date with recent Material
development and its support for more recent dependencies, e.g. TypeScript,
it is desirable for us to be on a newer version of Material.

This commit moves the Material tests away from the ivy-2019 branch, to a
recent commit on master. We are not targeting the master branch itself,
as that would introduce a moving target into Angular's CI checks, which
is undesirable.

Lastly, the usage of gulp to run Material's tests is changed into using
Bazel, as Material itself is now also built with Bazel.

JoostK commented on .circleci/config.yml:
```diff
@@ -654,6 +645,21 @@ jobs:
+          name: "Setup Bazel RBE remote execution for Material"
+          command: |
+            cd ${MATERIAL_REPO_TMP_DIR}
+            touch .bazelrc.user
```
Would love some input here. This block is mostly a copy of `setup_bazel_remote_execution `, but then targeting the Material repo. Note that `setup_bazel_remote_execution` is still part of this job's steps, as it initializes the GCP token (I don't think it's appropriate to use Material's key provisioning setup in the Angular repo).

Should we rework this into a script that can be used from `setup_bazel_remote_execution` and this step? Perhaps something different altogether?

devversion replied:
I could also see this as a separate script with an option to specify the path to the `bazelrc`. I think since we plan to keep the Material unit tests job for quite some time, it seems worth avoiding the duplication 🤷‍♂

josephperrott replied:
They are actually the same key as of now so it wouldn't be a big deal.  I have been thinking about potentially changing things so all three of our projects have different accounts and keys that are being used.

devversion commented on .circleci/config.yml:
```diff
@@ -620,19 +620,10 @@ jobs:
     steps:
       - *attach_workspace
       - *init_environment
+      - *setup_bazel_remote_execution
```
Can you add a comment that explains why we need this? (for the gcp token environment variable)

devversion commented on scripts/ci/run_angular_material_unit_tests.sh:
```diff
@@ -6,20 +6,28 @@ set -u -e -o pipefail
+# Now actually run the tests. The dev-app target is excluded as it fails to compile due to
+# limitations in Ivy's type checker (see FW-1352 and FW-1433)
+bazel build src/... --deleted_packages=//src/dev-app --build_tag_filters=-docs-package --define=compile=aot
+bazel test src/... --deleted_packages=//src/dev-app --build_tag_filters=-docs-package --define=compile=aot --test_tag_filters=-e2e
```
We could be a bit faster here by also excluding the e2e tag for the build targets. i.e. adding `,-e2e` to the `--build_tag_filters` for both commands.

devversion reviewed (approved):
LGTM.

josephperrott commented on scripts/ci/run_angular_material_unit_tests.sh:
```diff
@@ -6,20 +6,28 @@ set -u -e -o pipefail
+# Create a symlink for the Bazel binary installed through NPM, as running through Yarn introduces OOM errors.
+./scripts/circleci/setup_bazel_binary.sh
+
+# Now actually run the tests. The dev-app target is excluded as it fails to compile due to
```
We should note here that we are looking to both ensure that building and testing works.

If we were only doing testing then we wouldn't need the `yarn bazel build...` line.

JoostK replied:
@josephperrott Ah, I thought it was an optimization to run those separate. Am I correct in thinking that using just `bazel test` will still build most of the targets, as they are deps of the test targets? Or is there some other difference between `bazel build` vs `bazel test`?

josephperrott replied:
`bazel test` will "`bazel build`" everything it needs to execute the tests and then execute them.

In this case, yes `bazel test` will end up building most of the targets since we dep on them.

We chose to have both in the Components job so that we explicitly build everything and then test what we want to test explicitly.

JoostK replied:
Okay, then I think we should only run `bazel test` here

JoostK commented:
I notice now that all tests are run in both Chromium and Firefox, compared to only Chromium when we were using gulp. The total job duration has increased from ~3:30 to 7:30, which is quite a severe regression. @devversion Can we disable the Firefox targets using a filter? That should hopefully help to reduce the execution time.

josephperrott commented:
@JoostK We have a tag that you can target:

Example: https://github.com/angular/components/blob/master/package.json#L21 shows only targeting `chromium-local`

`bazel test //src/... --test_tag_filters=-e2e,-browser:firefox-local --build_tag_filters=-browser:firefox-local`

devversion commented:
@googlebot I consent.

gkalpak commented on .circleci/config.yml:
```diff
@@ -616,27 +621,17 @@ jobs:
-      - run:
-          name: "Cloning Material repository"
-          command: ./scripts/ci/clone_angular_material_repo.sh
       - restore_cache:
```
Why are we restoring the cache before cloning the repository?

devversion replied:
That was part of our discussion in the framework sync. We want to cache the repository since we always reset to a specific commit and don't want to re-clone the whole repository every time.

I'll try to add a comment here as well.

JoostK replied:
@devversion Can you comment on this and the remaining comments please?

gkalpak replied:
Yeah, sorry, it's obvious when you look at all changes. I left the comment before having looked through the whole PR and then forgot to remove it. +1 for adding a comment.

gkalpak commented on .circleci/config.yml:
```diff
@@ -616,27 +621,17 @@ jobs:
+          command: ./scripts/ci/clone_angular_material_repo.sh
       - save_cache:
-          # Material directory must be kept in sync with the `$MATERIAL_REPO_TMP_DIR` env variable.
-          # It needs to be hardcoded here, because env variables interpolation is not supported.
```
This comment is still relevant afaict.

devversion replied:
Re-added. good catch :smile:

gkalpak commented on .circleci/config.yml:
```diff
@@ -616,27 +621,17 @@ jobs:
+            - *material_unit_tests_cache_key_short
       - run:
-          name: Installing Material dependencies.
-          command: yarn --cwd ${MATERIAL_REPO_TMP_DIR} install --frozen-lockfile --non-interactive
```
So, where does this step happen now? 😕

devversion replied:
This step is not needed as the dependencies are installed through Bazel

gkalpak replied:
OOC, in which step are the dependencies added. Does this happen on the remote execution machines?

devversion replied:
It happens as part of a Bazel repository that is created through the `yarn_install` call in the `components` repository `WORKSPACE` file. That one creates the `@npm` repository by running `yarn`.

JoostK replied:
Sooo, actually this could only be removed due to the side-effects that `yarn add` has when updating to the `packages-dist` directory.

I replaced that call with a script that only updates the `package.json` file, without going though Yarn at all. This introduced a problem however, as we use Bazel itself from `node_modules` so we do in fact still require an initial yarn install (which was achieved using the `yarn add` call previously).

I readded this in a fixup cmmit as I don't see how we can do without this yarn install for now.

gkalpak reviewed (changes requested):
Looks promising :tada:
I left a few questions.

JoostK commented on .circleci/setup-rbe.sh:
```diff
@@ -0,0 +1,20 @@
+echo "Writing RBE configuration to ${bazelrc_user}"
+
+touch ${bazelrc_user}
+echo -e 'build --config=remote\n' >> ${bazelrc_user}
```
I looked into this last week, as I was interested to know this as well ;) I traced it back to https://github.com/angular/angular/commit/e8d0265c1e542976bbf54836bb5471e769cff5f3#diff-1d37e48f9ceff6d8030570cd36286a61, where it used to be echo'd into `/etc/bazel.bazelrc` which would probably have required elevated rights.

gkalpak replied:
> The echo -e is used so that the \n is interpreted as newline, AFAIK.

Doh! I was thinking that `echo` will add a `\n` at the end anyway and didn't realize the intention here is to have an extra newline :grin:

JoostK commented on .circleci/env.sh:
```diff
@@ -77,7 +77,7 @@ setPublicVar SAUCE_READY_FILE_TIMEOUT 120
 # their separate build setups.
 setPublicVar MATERIAL_REPO_TMP_DIR "/tmp/material2"
 setPublicVar MATERIAL_REPO_URL "https://github.com/angular/material2.git"
```
I decided not to do this at the moment, as it's all throughout the codebase (also filenames etc.).

devversion replied:
Yes, I saw this too but it's not in scope for this PR. Better in a follow-up :smile:

JoostK commented on scripts/ci/clone_angular_material_repo.sh:
```diff
@@ -6,5 +6,5 @@ set -u -e -o pipefail
 
 # Clone the Material repository into the given temporary directory.
-git clone --depth 1 --branch ${MATERIAL_REPO_BRANCH} ${MATERIAL_REPO_URL} \
+git clone --depth 1 --branch ${MATERIAL_REPO_COMMITSHA} ${MATERIAL_REPO_URL} \
```
This did not work, indeed. Paul's additional commit changed this right away. I can go back and fix the first commit, so that there's no confusion on this if someone were to look into the history of this file :)

gkalpak replied:
I would probably fix it, ut I don't feel strongly about it. Paul marked this conversation as resolved, so let's leave it as is.

devversion replied:
@gkalpak oh, I didn't realize that you requested updating the old commit. I'm fine either way :smile:

JoostK replied:
Fixed :-)

devversion commented on .circleci/env.sh:
```diff
@@ -77,7 +77,7 @@ setPublicVar SAUCE_READY_FILE_TIMEOUT 120
 setPublicVar MATERIAL_REPO_TMP_DIR "/tmp/material2"
 setPublicVar MATERIAL_REPO_URL "https://github.com/angular/material2.git"
-setPublicVar MATERIAL_REPO_BRANCH "ivy-2019"
+setPublicVar MATERIAL_REPO_COMMITSHA "036729dc1a7b41236fcda9b23c578f9ef65b484c"
```
We discussed this briefly in the components team meeting. The point was that we rather want to manually update the SHA instead of relying on a loose state of the `components` repository.

Not sure who is responsible for it though :smile:

JoostK commented on scripts/ci/run_angular_material_unit_tests.sh:
```diff
@@ -6,20 +6,27 @@ set -u -e -o pipefail
-# The angular/material2 CI sets TEST_PLATFORM to either "local", "saucelabs", or "browserstack".
-# For angular/angular, we only want to run the "local" tests.
-export TEST_PLATFORM=local
+yarn add file:${angular_dir}/dist/packages-dist-ivy-aot/{animations,bazel,common,compiler,compiler-cli,core,elements,forms,platform-server,platform-browser,platform-browser-dynamic,router,upgrade}
```
This is due to the introduction if Bazel. I don't know how Bazel's `yarn` related rules work, but I found it not to use the libraries as present in `node_modules`. Instead, it performs a seemingly clean install itself (to create the runfiles tree from) based on the `yarn.lock` file. That's why we're now explicitly installing the `packages-dist-ivy-aot` artifacts.

Maybe we can use a `find -f` query here and pass that to `yarn add`, but I'm no bash wizard 👼 <sup>stuff like `xargs` -> 🆘</sup>

gkalpak replied:
I am no bash wizard myself (and I always having a hard time bending `find` to my will :grin:). I think the following should work:

```
for pkg in ${angular_dir}/dist/packages-dist-ivy-aot/*; do yarn add file:$pkg; done
```

JoostK replied:
The problem with individual yarn adds is that it's slooooow :( I'll experiment a bit with `find`.

gkalpak replied:
Oh, right, you want them all in one string. How about:

```sh
yarn add file:${angular_dir}/dist/packages-dist-ivy-aot/{$(ls ${angular_dir}/dist/packages-dist-ivy-aot/ | xargs | tr ' ' ',')}
```

JoostK replied:
So it turns out that there's several packages within `packages-dist` that are not a dependency of Material, just adding them anyway sounds wrong. I decided to actually replace this with a script that carefully replaces `dependencies` and `devDependencies` in  the `package.json` file with references to the `dist-packages` directory.

It also prevents these warnings that were previously present in the job output:

```
warning "@angular/bazel" is already in "devDependencies". Please remove existing entry first before adding it to "dependencies".
```

After having applied this change it became apparent that the `yarn install` that Paul removed is still necessary, as we need Bazel from NPM in the first place. This was previously available as a side-effect of running `yarn add`, which performs an install by itself.

devversion committed:
build: lock material unit tests job to specific commit

No longer locks the Material unit tests job to a specific branch, but rather allows
locking to a specific commit from a given branch. This allows us to use the
"master" branch from the `components` repository.

JoostK committed:
fixup! build: lock material unit tests job to specific commit

gkalpak commented on scripts/ci/update-deps-to-dist-packages.js:
```diff
@@ -0,0 +1,58 @@
+/**
```
We have the same requirement (to replace `@angular/*` packages with locally built ones) for angular.io and docs examples tests. The "naive" approach of just replacing these packages in `package.json` eventually breaks when one of the peer dependencies is not satisfied any more (e.g. the min required TypeScript version).

To avoid this issue, we have built [this script][1] for `aio/`. Maybe we could re-use that?

[1]: https://github.com/angular/angular/blob/c7542a1d09b27b027a0367a1f3732a908cdd617a/aio/tools/ng-packages-installer/index.js

JoostK replied:
Discussed offline that we can do without the AIO script.

JoostK commented on .circleci/config.yml:
```diff
@@ -642,6 +642,10 @@ jobs:
+      - run:
+          # Run yarn install to fetch the Bazel binaries as used in the Material repo.
+          name: Installing Material dependencies.
+          command: yarn --cwd ${MATERIAL_REPO_TMP_DIR} install --frozen-lockfile --non-interactive
```
I inserted it back into this place as it used to be here originally. I don't know what the benefits of caching `node_modules` would be, but I assume it just results in a larger cache (`node_modules` can be large!) and given that we run with `--frozen-lockfile`, Yarn does not have to perform resolution but just download all packages. And even when caching we need to run `yarn install`, as the cache might get out of date whenever the commit sha is updated. @devversion Do you think we should include `node_modules` in the cache, or not?

devversion replied:
hmm that is a good question. When I did the changes for the new cloning w/ specific SHA, I removed this `yarn install` step, but now since we added it back I feel like we still should cache the node modules too. It still feels a bit weird to me that we need to pull in the whole node modules a second time just in favor of bringing in Bazel... wondering if we should rethink which Bazel version we should use? an option would be to use the `bazel` version from the `framework` node modules (which could cause version conflicts..)

we could also just include the whole `node_modules` in the cached repository (sounds like the easiest and most reliable solution for now)

JoostK replied:
This step is now cached, and it has improved job run time by ~30-35s (CCI is slow right now so it's a bit hard to tell). The cache size increased from 30M to 200M, which is quite decent to me.

gkalpak replied:
Good call caching `node_modules/` :wink:
We do this for other jobs (so why not here) and one of the main motivations was avoiding flakes due to yarn timeouts when fetching packages, especially large ones (such as bazel binaries :wink:).

JoostK committed:
fixup! test(ivy): update Material to recent commit from master branch

JoostK committed:
fixup! build: lock material unit tests job to specific commit

JoostK commented:
We typically target `master & patch` for build related PRs, however the Material tests might not run correctly on the patch version (which is lacking Ivy fixes). So regarding the target, I propose this is targeting only `master` and I will open another PR targeting `patch` to see if it's green and can be merged into `patch`.

gkalpak reviewed (approved):
:tada:
